File: javascript/cote/LTCD/순열.mjs

const TEST_CASES = 10
const N = 8
const MAX_VALUE = 100


function solve(nums) {
	const ret = []

	function dfs(components) {
		if (components.length == 0) {
			ret.push([...ret])
			return
		}

		for (let component of components) {
			ret.push(component)
			dfs(components.filter(v => v != component))
			ret.pop()
		}
	}

	dfs([...nums])

	return ret
}


for (let it =0; it < TEST_CASES; it++) {
	let nums = []
	let n = Math.floor(Math.random() * (N - 1)) + 1
	
	for (let i = 0; i < n; i++) {
		nums[i] = Math.floor(Math.random() * MAX_VALUE)
	}
	console.log(nums.length)
	console.log(solve(nums))
}

File: python/cote/LTCD/순열.py
from typing import List
import itertools
import random

TEST_CASES = 10
N = 5


def solve(nums: List[int]) -> List[List[int]]:
    ret = []
    prev = []

    def dfs(elements: List[int]):

        if len(elements) == 0:
            ret.append(prev[:])

        for e in elements:
            next = elements[:]
            next.remove(e)

            prev.append(e)
            dfs(next)
            prev.pop()

    dfs(nums)
    return ret


def solve2(nums: List[int]) -> List[List[int]]:
    return list(itertools.permutations(nums))


if __name__ == "__main__":
    random.seed(43)

    for _ in range(TEST_CASES):
        nums = random.sample(range(1, 3 * N), N)
        print(solve(nums), solve2(nums))
